packages/hash/shared/src/ProsemirrorManager.ts

```diff
@@ -0,0 +1,530 @@
+        ? entityStore?.draft[draftBlockId]
+        : null;
+
+    const childDraftId = isDraftBlockEntity(blockEntity)
+      ? blockEntity.properties.entity.draftId
+      : null;
+
+    const blockData =
+      draftBlockId && entityStore
+        ? getBlockChildEntity(draftBlockId, entityStore)
+        : null;
+
+    const content =
+      blockData && isTextEntity(blockData)
+        ? childrenForTextEntity(blockData, this.schema)
+        : [];
+
+    return this.schema.nodes.block!.create({}, [
+      this.schema.nodes.entity!.create(
+        { draftId: draftBlockId },
+        this.schema.nodes.entity!.create(
+          {
+            draftId: childDraftId,
+          },
+          [this.schema.nodes[targetComponentId]!.create({}, content)],
+        ),
+      ),
+    ]);
+  }
+
+  private assertBlockDefined(componentId: string) {
+    if (!this.schema.nodes[componentId]) {
+      throw new Error("Block must already be defined");
+    }
+  }
+
+  async loadPage(currentState: EditorState<Schema>, entities: BlockEntity[]) {
+    const store = createEntityStore(
+      entities,
+      entityStorePluginState(currentState).store.draft,
+    );
+
+    const newNodes = await Promise.all(
+      entities.map(async (blockEntity) => {
+        const draftEntity = mustGetDraftEntityByEntityId(
+          store.draft,
+          blockEntity.entityId,
+        );
+
+        await this.defineBlockByComponentId(blockEntity.properties.componentId);
+
+        return this.renderBlock(
+          blockEntity.properties.componentId,
+          store,
+          draftEntity.draftId,
+        );
+      }),
+    );
+
+    const { tr } = currentState;
+
+    addEntityStoreAction(currentState, tr, { type: "store", payload: store });
+
+    tr.replaceWith(0, currentState.doc.content.size, newNodes);
+
+    return tr;
+  }
+
+  /**
+   * @todo consider removing the old block from the entity store
+   */
+  async replaceNode(
+    draftBlockId: string,
+    targetComponentId: string,
+    targetVariant: BlockVariant,
+    node: ProsemirrorNode<Schema>,
+    pos: number,
+  ) {
+    const { view } = this;
+
+    // @todo consider separating the logic that uses a view to something else
+    if (!view) {
+      throw new Error("Cannot trigger replaceNodeWithRemoteBlock without view");
+    }
+
+    const [tr, newNode] = await this.createBlock(
+      targetComponentId,
+      draftBlockId,
+      targetVariant,
+    );
+
+    tr.replaceRangeWith(pos, pos + node.nodeSize, newNode);
+    view.dispatch(tr);
+  }
+
+  /**
+   * @todo consider removing the old block from the entity store
+   */
+  async deleteNode(node: ProsemirrorNode<Schema>, pos: number) {
+    const { view } = this;
+
+    if (!view) {
+      throw new Error("Cannot trigger deleteNode without view");
+    }
+
+    const { tr } = view.state;
+
+    tr.delete(pos, pos + node.nodeSize);
+    view.dispatch(tr);
+  }
+
+  async createBlock(
```

Comment:
It feels this method also handles updating the block/swapping. If we can't rename, then we should probably add a comment explaining what it does

Reply:
This never updates a block itself. It can update block data.

The purpose of this block is to render a component specified by a `targetComponentId` with a block specified by a `draftBlockId`. It may update the block data if a specific variant of the component is specified.

Mostly it just works out if the `draftBlockId` and `targetComponentId` are compatible, and creates a new block if not.